compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/TypeDefBuilderHelper.java

```diff
@@ -299,4 +306,67 @@ public static String getPackageAlias(SymbolEnv env, String compUnitName, Package
+        BTypeSymbol immutableStructureSymbol = structureType.tsymbol;
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+        for (BField origField : origStructureType.fields.values()) {
+            BType fieldType;
+            if (isImmutable) {
+                fieldType = ImmutableTypeCloner.getImmutableType(pos, types, origField.type, env,
+                        env.enclPkg.packageID, env.scope.owner, symTable, anonymousModelHelper, names, unresolvedTypes);
+            } else {
+                fieldType = origField.type;
+            }
+
+            Name origFieldName = origField.name;
+            BVarSymbol immutableFieldSymbol;
+            if (fieldType.tag == TypeTags.INVOKABLE && fieldType.tsymbol != null) {
+                immutableFieldSymbol = new BInvokableSymbol(origField.symbol.tag, origField.symbol.flags | flag,
+                        origFieldName, pkgID, fieldType,
+                        immutableStructureSymbol, origField.symbol.pos, SOURCE);
+                BInvokableTypeSymbol tsymbol = (BInvokableTypeSymbol) fieldType.tsymbol;
+                BInvokableSymbol invokableSymbol = (BInvokableSymbol) immutableFieldSymbol;
+                invokableSymbol.params = tsymbol.params == null ? null : new ArrayList<>(tsymbol.params);
+                invokableSymbol.restParam = tsymbol.restParam;
+                invokableSymbol.retType = tsymbol.returnType;
+                invokableSymbol.flags = tsymbol.flags;
+            } else if (fieldType == symTable.semanticError) {
+                // Can only happen for records.
+                immutableFieldSymbol = new BVarSymbol(origField.symbol.flags | flag | Flags.OPTIONAL,
+                        origFieldName, pkgID, symTable.neverType,
+                        immutableStructureSymbol, origField.symbol.pos, SOURCE);
+            } else {
+                immutableFieldSymbol = new BVarSymbol(origField.symbol.flags | flag, origFieldName, pkgID,
+                        fieldType, immutableStructureSymbol,
+                        origField.symbol.pos, SOURCE);
+            }
+            String nameString = origFieldName.value;
+            fields.put(nameString, new BField(origFieldName, null, immutableFieldSymbol));
+            immutableStructureSymbol.scope.define(origFieldName, immutableFieldSymbol);
+        }
+        structureType.fields = fields;
+
+        if (origStructureType.tag == TypeTags.OBJECT) {
+            return;
+        }
+
+        BLangUserDefinedType origTypeRef = new BLangUserDefinedType(
+                ASTBuilderUtil.createIdentifier(pos,
+                        TypeDefBuilderHelper.getPackageAlias(env, pos.lineRange().filePath(),
+                                origStructureType.tsymbol.pkgID)),
+                ASTBuilderUtil.createIdentifier(pos, origStructureType.tsymbol.name.value));
+        origTypeRef.pos = pos;
+        origTypeRef.setBType(origStructureType);
```

Comment:
Also needs to be done only if `isImmutable` is true? IMO, we can move this logic also to where `populateStructureFields` is called since it is only needed in the immutable case.

Reply:
There are some failing tests with this change. Shall I revert the changes to originals?